### ISSUE TITLE
Generics extending type do not enforce type consistantly

### ISSUE THREAD

**Issue by ycmjason** [microsoft/TypeScript#31403]
**TypeScript Version:** 
3.4

**Search Terms:** 
generics, type, extends

**Code and Playground Link:** 
[playground 1](https://www.typescriptlang.org/play/#src=type%20Person%20%3D%20%7B%20name%3A%20string%2C%20age%3A%20number%20%7D%3B%0D%0Atype%20PeopleMap%20%3D%20%7B%20%5Bk%3A%20string%5D%3A%20Person%20%7D%3B%0D%0A%0D%0Aconst%20t%20%3D%20(%3CT%20extends%20PeopleMap%3E(o%3A%20T)%20%3D%3E%20o)(%7B%0D%0A%20%20jason%3A%20%7B%20name%3A%20'jason'%2C%20age%3A%203%2C%20k%3A%203%20%7D%2C%0D%0A%20%20jane%3A%20%7B%20name%3A%20'jane'%2C%20age%3A%203%20%7D%2C%0D%0A%7D)%3B) (happy ts, sad developer)
```ts
type Person = { name: string, age: number };
type PeopleMap = { [k: string]: Person };

const t = (<T extends PeopleMap>(o: T) => o)({
  jason: { name: 'jason', age: 3, k: 3 }, // does not complain about k: 3
  jane: { name: 'jane', age: 3 },
});
```

[playground 2](https://www.typescriptlang.org/play/#src=type%20Person%20%3D%20%7B%20name%3A%20string%2C%20age%3A%20number%20%7D%3B%0D%0Atype%20PeopleMap%20%3D%20%7B%20%5Bk%3A%20string%5D%3A%20Person%20%7D%3B%0D%0A%0D%0Aconst%20t%20%3D%20(%3CT%20extends%20PeopleMap%3E(o%3A%20T)%20%3D%3E%20o)(%7B%0D%0A%20%20jason%3A%20%7B%20name%3A%20'jason'%2C%20age%3A%203%2C%20k%3A%203%20%7D%2C%0D%0A%20%20jane%3A%20%7B%20name%3A%20'jane'%2C%20age%3A%203%20%7D%2C%0D%0A%20%20mary%3A%20%7B%20not%3A%203%20%7D%2C%0D%0A%7D)%3B) (sad ts, happy developer)
```ts
type Person = { name: string, age: number };
type PeopleMap = { [k: string]: Person };

const t = (<T extends PeopleMap>(o: T) => o)({
  jason: { name: 'jason', age: 3, k: 3 }, // error at `k: 3`
  jane: { name: 'jane', age: 3 },
  mary: { not: 3 }, // error at `not: 3`
});
```

**Expected behavior:**
I expect TS to complain in both scenario. 

- Scenario 1 should complain about `k: 3` being not assignable to `Person`.
- Scenario 2 is complaining correctly as expected.

**Actual behavior:**
- Scenario 1 is not complaining!

**Related Issues:** 
Not that I found.

**Comment by dragomirtitian** [comment]
Very similar to #29519. This is working as intended. 

When inferring `T`, the constraint specifies a minimal set of properties `T` must have. But `T` can have more properties (it can be a subtype of the constraint). This also applies to the properties of `T`. Even though the constraint of `T` specifies `Person`, `T` can have properties that are a subtype of `Person`. So not getting an error is the desired behavior in the first example.

When inference fails (such as in your second example, where it fails since the object literal is not a subtype of the constraint `PeopleMap`) the fallback is to use the constraint as `T` and continue checking. This means you will get the errors on why the `mary` is not a `Person`, but it will also mean excess property checks kick in and you get the error on `k`.

@RyanCavanaugh  maybe excess property checks can be disabled when inference fails as they can't be the real source of the error as far as I can tell?

**FYI:** If you want to disallow `k` you can use a custom type to trigger an error on any extra properties:

```ts
type Person = { name: string, age: number };
type PeopleMap = { [k: string]: Person };

type ConstrainPropertie<T, C extends Record<string, unknown>> = {
    [P in keyof T]: Record<Exclude<keyof T[P], keyof C[string]>, never>
}
const fn = (<T extends PeopleMap>(o: T & ConstrainPropertie<T, PeopleMap>) => o)
const t = fn({
    jason: { name: 'jason', age: 3, k: 3 },
    jane: { name: 'jane', age: 3 },
});
```
A bit ugly but don't think there is a better option.

**Comment by ycmjason** [comment]
Thank you for your reply. It is very informative.

Please don't mind my follow-up question.

```
type Person = { name: string, age: number };
const people = {
    jason: { name: 'jason', age: 3 },
    jane: { name: 'jane', age: 3 },
}

people.jason;
```

If I wish to make each value in `people` to be of type `Person`, is there a way to do it so that TS can provide type autocomplete/error? It is also required that clicking on `jason` in the IDE will jump to its definition. (So it can't be `Record<string, Person>`?)

**Comment by dragomirtitian** [comment]
@ycmjason Your approach of using a function is the way I usually recommend doing this as well. I don't know of a better way doing it since for a variable you can't at tell the compiler to infer the value but also specify a constraint, only functions can do that.

**Comment by ycmjason** [comment]
Understood. So I suppose the method you suggested is the best way to do it for now. 

Thank you.

**Comment by RyanCavanaugh** [comment]
> @RyanCavanaugh maybe excess property checks can be disabled when inference fails as they can't be the real source of the error as far as I can tell?

Perhaps, but it's hard from an implementation perspective. What really happens when inference "fails" is that we pick a candidate anyway, then proceed with the call knowing (well, predicting) that one of the arguments will not be assignable to its parameter type. There'd have to be a bunch of trafficking around of extra information to make it happen.
